### ISSUE TITLE
A&R: Spreadsheet template idea using CDOGS

### ISSUE THREAD

**Issue by Dianadec** [bcgov/bcparks-ar-admin#386]
#### Description:

This ticket is to explore the use of [CDOGS](https://bcgov.github.io/common-service-showcase/services/cdogs.html) to support A&R's use of spreadsheets in an easier way. 

> "Leverage your structured datasets and your business templates to automatically populate printable documents, spreadsheets, presentations, or PDFs using the Common Document Generation Service."

#### Acceptance Criteria:

Given: I am the user
When: I request an export of the data
Then: I am able to receive the data in a template I recognize, so I don't have to further manipulate the data

Given: I am the developer
When: I create a new export (using a template given to me from the user)
Then: I can send the data to CDOGS to generate a document in a format the user is familiar with

#### Development Checklist:

- [ ] ...
- [ ] ...
- [ ] ...

#### Dependencies

- Blocked by
- Blocking

**Relevant documentation as reference**

- Example:
  - Template:
[template_movies.xlsx](https://github.com/user-attachments/files/17154301/template_movies.xlsx)
  - Data passed in:
[contexts_movies.json](https://github.com/user-attachments/files/17154306/contexts_movies.json)
  - Output: [movies_output.xlsx](https://github.com/user-attachments/files/17154313/movies_output.xlsx)

This is an internal BC Gov tool. For example, the [WALLY](https://github.com/bcgov/WALLY/) app used this tool to export datasets for watersheds. It provides a bit more flexibility for displaying the data in an excel spreadsheet for the user, allowing some additional formatting, colour-coding, and allows multiple sheets to be added.

The user creates a [template .xlsx file](https://github.com/bcgov/WALLY/blob/4c3866d0fdb41be826bf00390ad05a792f85b6fb/backend/api/v1/watersheds/templates/SurfaceWater.xlsx) and then makes an [API call to CDOGS](https://github.com/bcgov/WALLY/blob/4c3866d0fdb41be826bf00390ad05a792f85b6fb/backend/external/docgen/controller.py#L19) with the [data in JSON format](https://github.com/bcgov/WALLY/blob/4c3866d0fdb41be826bf00390ad05a792f85b6fb/backend/api/v1/watersheds/routes.py#L23) that will fit the template.

**Definition of Ready**

- [ ] Acceptance criteria are included
- [ ] Wireframes are included (if applicable)
- [ ] Design / Solution is accepted by Product Owner (if applicable)
- [ ] Dependencies are identified (technical, business, regulatory/policy)
- [ ] Story has been estimated (under 13 pts)

**Definition of Done** 
- In progress:
  - [ ] Acceptance criteria are tested (Functionality meets the acceptance criteria defined in the ticket)
  - [ ] UI meets accessibility requirements
  - [ ] Unit tests are written
  - [ ] Work is traceable in GitHub
  - [ ] PR linked to ticket number
  - [ ] If needed/required - Dev adds flag/label to highlight any migration steps necessary prior to PROD deployment
- Code review:
  - [ ] Code is peer reviewed and has passed CI/CD tests
- QA:
  - [ ] Acceptance criteria are tested (Functionality meets the acceptance criteria defined in the ticket)
  - [ ] Code is potentially shippable to the production environment
  - [ ] Functional features have been tested and passed by QA
  - [ ] UI components tested by designer
  - [ ] Code is deployed to PROD when moved to 'done' column (unless requested otherwise by PO)
- PO Review:
  - [ ] Acceptance criteria are tested (Functionality meets the acceptance criteria defined in the ticket)
  - [ ] Reviewed and approved by Product Owner

#### Notes:
- 
-

**Comment by Dianadec** [comment]
@davidclaveau here's the ticket I made - feel free to add more info as needed :) thanks!

**Comment by davidclaveau** [comment]
# Findings!

## TL;DR:

CDOGS might be useful to implement in future export tickets, but overall it requires probably about the same amount of work as the current CSV export. The current export setup is great for getting all the data into the user's hands. However, if a user frequently needs to create a complex, standardized - and time-consuming - spreadsheet from exported data, and this spreadsheet follows a consistent structure each time it's generated, then CDOGS would probably be a useful timesaving option.

Next time we have a request to create an export and new spreadsheet for A&R, we should look at this again to see if CDOGS would be a suitable fit.

## Pros and Cons:

### Pros:
- A user can more easily format and style a spreadsheet to their liking. Makes it easier for the user to read and parse from the get-go.
	- e.g colour-coded sections, bolded titles, properly spaced columns and rows, etc.
- A user can have several sheets to accommodate data, so they're not limited to one sheet with a big data dump
	- e.g. the user could separate Frontcountry, Backcountry, Boating, or Day Use data to different sheets
- It's (*maybe*) a be a bit easier for the developer to get the data from the backend onto the spreadsheet, although it requires learning a way to drill to this data.
	- The real timesaver is that you essentially only need to build the data object. You don't need to then take the data object and do weird looping steps to add commas and separate the data into a valid CSV item.
- There is some functionality that Carbone (what CDOGS uses under the hood) that makes it quite useful, like conditional logic and filtering.
	- This might be easier to do on the backend anyway, as we're doing data manipulation to build the data object anyway.
- Can do other documents besides spreadsheets 

### Cons:
- Can't use existing Excel functions for data manipulation, so we can't add items to cell A1 and B1 and then use `=SUM(A1,B1)` as this doesn't work.
	- There seem to [workarounds](https://help.carbone.io/en-us/article/how-to-compute-totals-in-excel-spreadsheets-when-carbone-injects-new-lines-1iat7yy/), but I couldn't get this to work in my template for some reason. Might require a bit more fiddling, but it's possible we don't have the most up to date version of Carbone. 
- Learning a new syntax for getting the data into the template, so takes some time to manipulate and drill into objects in the template.
	- Still requires a bit of data manipulation on the backend to pull all the data needed and create a data object that can be passed to the template to dynamically build the spreadsheet.

## More robust summary:

The main purpose of this ticket was to investigate how useful it would be to integrate CDOGS for A&R exports. In practice, an A&R user would provide a template spreadsheet (or other document) to the developer, who would then modify it to create reusable, dynamic reports for future exports following that same template design. 

For a general export of all the data (like the data we get from the current Standard or Variance reports), integrating CDOGS would probably be about the same amount of work. You still need to create an export function in the backend, make the necessary queries to the database to retrieve the data, and then manipulate it to build the data object. With CDOGS, you then need to use Carbone's syntax in the Excel template to populate the cells from the data object. Comparatively, with the current setup we use, it builds a new, comma-separated data object (the CSV), so you only really save time in not having to do loops in the code to build the CSV object (and even then it's not that much time).

Some features were also missing from CDOGS that would have been useful. For example, I initially assumed that the native features in Excel could be utilized in the template, such as utilizing existing Excel functions (like `=SUM()` or `=IF()`) to show where/how data should be manipulated. But these functions cannot run on "dynamically populated data" when the spreadsheet is built. There were workarounds but I couldn't get them to work (which might be a versioning issue).

Like other programming languages, Carbone offers some of its own syntax to manipulate the data right in the template, such as conditional logic like `if/else`. It also has some nice-to-have formatting syntax, like date conversions, appends, prepends, string manipulation (e.g. change a text field to uppercase), and so on - so you can do quite a bit to the data as it's populating the template. However, it would probably make more sense to manipulate these items in the backend when exporting the data.
